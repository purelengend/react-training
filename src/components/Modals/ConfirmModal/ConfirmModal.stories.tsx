import type { Meta, StoryObj } from '@storybook/react';
import ConfirmModal from '@components/Modals/ConfirmModal';
import '@components/Modals/ConfirmModal/confirm-modal.module.css';
const meta: Meta<typeof ConfirmModal> = {
  title: 'Example/ConfirmModal',
  component: ConfirmModal,
  decorators: [
    Story => (
      <div
        style={{
          display: 'grid',
          placeItems: 'center',
          height: '90vh'
        }}
      >
        <Story />
      </div>
    )
  ],
  argTypes: {
    onSubmit: {
      description: 'The submit function to be called when submitting the modal',
      action: 'submitted'
    },
    dataId: {
      description: 'The value of hidden id input field of the modal',
      defaultValue: '0'
    },
    message: {
      description: 'The message of the modal'
    },
    onCancelClick: {
      description: 'The cancel method is used to close the modal',
      action: 'closed'
    }
  },
  parameters: {
    controls: {
      expanded: true
    }
  }
};

type Story = StoryObj<typeof ConfirmModal>;

export const ConfirmModalExample: Story = {
  args: {
    message: 'Sample message of the modal',
    onSubmit: e => {
      e.preventDefault();
    }
  }
};
export default meta;
