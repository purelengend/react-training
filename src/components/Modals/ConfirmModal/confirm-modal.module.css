.confirm-modal-body {
  max-width: 480px;
  padding: 32px;
  border-radius: var(--border-secondary);
  background-color: var(--bg-secondary);
  gap: 48px;
  justify-content: space-between;
}

.confirm-modal-content {
  font-size: var(--font-size-xl);
  color: var(--text-primary);
  font-weight: var(--font-weight-bold);
  line-height: var(--line-height-lg);
  text-align: center;
}

.confirm-modal-btn-wrapper {
  justify-content: space-between;
}
