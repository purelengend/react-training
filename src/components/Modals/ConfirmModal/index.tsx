import { FormEvent, memo } from 'react';
import { Button } from '../../common/Button';
import confirmModalStyles from './confirm-modal.module.css';
interface ConfirmModalProps {
  isVisible: boolean;
  message: string;
  onSubmit: (event: FormEvent<HTMLFormElement>) => void;
}
const ConfirmModal = memo(
  ({ isVisible, message, onSubmit }: ConfirmModalProps) => {
    return (
      isVisible && (
        <div id="confirm-modal" className="d-flex-center modal-overlay">
          <form
            id="confirm-form"
            className={`d-flex-col ${confirmModalStyles['confirm-modal-body']}`}
            onSubmit={onSubmit}
          >
            <input
              id="hidden-confirm-id"
              type="number"
              defaultValue="0"
              name="id"
              hidden
            />
            <h2 className={confirmModalStyles['confirm-modal-content']}>
              {message}
            </h2>
            <div
              className={`d-flex ${confirmModalStyles['confirm-modal-btn-wrapper']}`}
            >
              <Button className="modal-btn cancel">Cancel</Button>
              <Button type="submit" className="modal-btn confirm">
                Yes
              </Button>
            </div>
          </form>
        </div>
      )
    );
  }
);

export default ConfirmModal;
