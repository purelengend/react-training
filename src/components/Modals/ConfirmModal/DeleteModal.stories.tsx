import type { Meta, StoryObj } from '@storybook/react';
import ConfirmModal from '@components/Modals/ConfirmModal';
import '@components/Modals/ConfirmModal/confirm-modal.module.css';
const meta: Meta<typeof ConfirmModal> = {
  title: 'Example/ConfirmModal',
  component: ConfirmModal,
  decorators: [
    Story => (
      <div
        style={{
          display: 'grid',
          placeItems: 'center',
          height: '90vh'
        }}
      >
        <Story />
      </div>
    )
  ],
  argTypes: {
    isVisible: {
      description: 'Indicates whether the modal is hidden or visible.'
    },
    onSubmit: {
      description: 'The submit function to be called when submitting the modal',
      action: true
    },
    dataId: {
      description: 'The value of hidden id input field of the modal',
      defaultValue: '0'
    },
    message: {
      description: 'The message of the modal'
    }
  },
  parameters: {
    controls: {
      expanded: true
    }
  },
  
};

type Story = StoryObj<typeof ConfirmModal>;

export const ConfirmModalExample: Story = {
  args: {
    isVisible: true,
    message: 'Sample message of the modal',
    onSubmit: e => {
      e.preventDefault();
    }
  }
};
export default meta;
