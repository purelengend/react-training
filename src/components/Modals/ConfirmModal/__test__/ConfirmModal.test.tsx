import { customRenderer, render, screen } from '@src/test/test-utils';
import userEvent from '@testing-library/user-event';
import { FormEvent } from 'react';

import ConfirmModal, { ConfirmModalProps } from '..';

describe('ConfirmModal test case', () => {
  const mockConfirmModalProps: ConfirmModalProps = {
    message: 'mock',
    onCancelClick: vi.fn(),
    onSubmit: vi
      .fn()
      .mockImplementation((e: FormEvent<HTMLFormElement>) => e.preventDefault())
  };

  it('should render correctly', () => {
    const mockConfirmModal = customRenderer(
      <ConfirmModal {...mockConfirmModalProps} />
    );

    expect(mockConfirmModal.toJSON()).toMatchSnapshot();
  });

  it('should invoke onSubmit function when submitting', async () => {
    render(<ConfirmModal {...mockConfirmModalProps} />);

    const mockButtonList = screen.getAllByRole('button');

    await userEvent.click(mockButtonList[1]);

    expect(mockConfirmModalProps.onSubmit).toBeCalled();
  });

  it('should invoke onCancel function when closing', async () => {
    render(<ConfirmModal {...mockConfirmModalProps} />);

    const mockButtonList = screen.getAllByRole('button');

    await userEvent.click(mockButtonList[0]);

    expect(mockConfirmModalProps.onCancelClick).toBeCalled();
  });
});
