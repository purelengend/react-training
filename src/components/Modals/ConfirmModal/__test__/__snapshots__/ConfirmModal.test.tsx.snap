// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmModal test case > should render correctly 1`] = `
<div
  className="d-flex-center modal-overlay"
  id="confirm-modal"
>
  <form
    className="d-flex-col _confirm-modal-body_1874c6"
    id="confirm-form"
    onSubmit={[MockFunction spy]}
  >
    <h2
      className="_confirm-modal-content_1874c6"
    >
      mock
    </h2>
    <div
      className="d-flex _confirm-modal-btn-wrapper_1874c6"
    >
      <button
        className="modal-btn cancel"
        data-id="0"
        disabled={false}
        onClick={[MockFunction spy]}
        type="button"
      >
        Cancel
      </button>
      <button
        className="modal-btn confirm"
        data-id="0"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Yes
      </button>
    </div>
  </form>
</div>
`;
